### ISSUE TITLE
Index.html

### ISSUE THREAD

**Issue by pubgadda0071-ctrl** [pubgadda0071-ctrl/Index.html#1]
"""
rsi_gui_app_fixed.py
Tkinter + mplfinance app that fetches Binance klines, computes RSI, shows candlesticks + RSI,
and displays BUY/SELL/HOLD signal based on RSI bands (75/50/30).

Run: python rsi_gui_app_fixed.py
"""

import tkinter as tk
from tkinter import ttk, messagebox
import threading
import time
import requests
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib
matplotlib.use("TkAgg")
import matplotlib.pyplot as plt
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import mplfinance as mpf
import queue

# ----------------- CONFIG -----------------
BINANCE_KLINES = "https://api.binance.com/api/v3/klines"
SYMBOLS = {
    "Bitcoin (BTC/USDT)"  : "BTCUSDT",
    "Ethereum (ETH/USDT)" : "ETHUSDT",
    "BNB (BNB/USDT)"      : "BNBUSDT",
    "Solana (SOL/USDT)"   : "SOLUSDT",
    "Cardano (ADA/USDT)"  : "ADAUSDT",
    "Dogecoin (DOGE/USDT)": "DOGEUSDT"
}
DEFAULT_INTERVAL = "1m"
CANDLES_LIMIT = 200
DEFAULT_REFRESH = 10  # seconds
RSI_PERIOD = 14
UPPER_BAND = 75
MIDDLE_BAND = 50
LOWER_BAND = 30
MAX_HISTORY = 200

# ----------------- Utilities -----------------
def fetch_klines(symbol: str, interval: str = DEFAULT_INTERVAL, limit: int = CANDLES_LIMIT):
    """Fetch klines from Binance and return DataFrame with Open/High/Low/Close/Volume indexed by datetime."""
    params = {"symbol": symbol, "interval": interval, "limit": limit}
    resp = requests.get(BINANCE_KLINES, params=params, timeout=10)
    resp.raise_for_status()
    data = resp.json()
    df = pd.DataFrame(data, columns=[
        "open_time", "Open", "High", "Low", "Close", "Volume",
        "close_time","q","n","taker_base_vol","taker_quote_vol","ignore"
    ])
    df["Open"]  = df["Open"].astype(float)
    df["High"]  = df["High"].astype(float)
    df["Low"]   = df["Low"].astype(float)
    df["Close"] = df["Close"].astype(float)
    df["Volume"]= df["Volume"].astype(float)
    df["Date"]  = pd.to_datetime(df["open_time"], unit="ms")
    df.set_index("Date", inplace=True)
    return df[["Open","High","Low","Close","Volume"]]

def compute_rsi(series: pd.Series, period: int = RSI_PERIOD):
    """Compute RSI using Wilder's smoothing (EMA)."""
    delta = series.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.ewm(alpha=1/period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1/period, adjust=False).mean()
    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))
    return rsi

def format_price(p):
    """Format price trimming unnecessary zeros."""
    s = f"{p:,.8f}"
    # remove trailing zeros and dot if integer-like
    s = s.rstrip("0").rstrip(".")
    return s

# ----------------- Background fetcher thread -----------------
class FetchThread(threading.Thread):
    def __init__(self, symbol_getter, interval_getter, refresh_getter, out_queue, stop_event):
        super().__init__(daemon=True)
        self.symbol_getter = symbol_getter
        self.interval_getter = interval_getter
        self.refresh_getter = refresh_getter
        self.out_queue = out_queue
        self.stop_event = stop_event

    def run(self):
        while not self.stop_event.is_set():
            try:
                symbol_key = self.symbol_getter()
                symbol = SYMBOLS.get(symbol_key, list(SYMBOLS.values())[0])
                interval = self.interval_getter() or DEFAULT_INTERVAL
                df = fetch_klines(symbol, interval=interval, limit=CANDLES_LIMIT)
                if df.empty:
                    raise RuntimeError("No data returned")
                df["RSI"] = compute_rsi(df["Close"], period=RSI_PERIOD)
                latest_price = float(df["Close"].iloc[-1])
                latest_rsi = df["RSI"].iloc[-1]
                payload = {"symbol": symbol, "df": df, "price": latest_price, "rsi": float(latest_rsi)}
                self.out_queue.put(("data", payload))
            except Exception as e:
                self.out_queue.put(("error", str(e)))
            # Sleep using refresh getter value (in seconds)
            try:
                secs = int(self.refresh_getter())
                if secs < 1:
                    secs = DEFAULT_REFRESH
            except:
                secs = DEFAULT_REFRESH
            # Wait but exit early if stop_event set
            for _ in range(secs):
                if self.stop_event.is_set():
                    break
                time.sleep(1)

# ----------------- Main App -----------------
class RSIApp:
    def __init__(self, root):
        self.root = root
        self.root.title("RSI Crypto Live — Fixed")
        self.root.geometry("1100x700")

        # UI/State
        self.history = []
        self.queue = queue.Queue()
        self.stop_event = threading.Event()
        self.fetch_thread = None

        # Controls frame
        ctrl = tk.Frame(root)
        ctrl.pack(fill="x", padx=8, pady=6)

        tk.Label(ctrl, text="Symbol:", font=("Segoe UI", 10)).pack(side="left")
        self.combo = ttk.Combobox(ctrl, values=list(SYMBOLS.keys()), state="readonly", width=30)
        self.combo.current(0)
        self.combo.pack(side="left", padx=6)

        tk.Label(ctrl, text="Interval:", font=("Segoe UI", 10)).pack(side="left", padx=(10,0))
        self.interval = ttk.Combobox(ctrl, values=["1m","3m","5m","15m","1h","4h"], width=6)
        self.interval.set(DEFAULT_INTERVAL)
        self.interval.pack(side="left", padx=6)

        tk.Label(ctrl, text="Refresh (s):", font=("Segoe UI", 10)).pack(side="left", padx=(10,0))
        self.ref_entry = tk.Entry(ctrl, width=5)
        self.ref_entry.insert(0, str(DEFAULT_REFRESH))
        self.ref_entry.pack(side="left", padx=6)

        self.btn_start = tk.Button(ctrl, text="Start", command=self.start_fetch, bg="#2e7d32", fg="white")
        self.btn_start.pack(side="left", padx=6)
        self.btn_stop = tk.Button(ctrl, text="Stop", command=self.stop_fetch, bg="#b71c1c", fg="white", state="disabled")
        self.btn_stop.pack(side="left", padx=6)
        self.btn_manual = tk.Button(ctrl, text="Refresh Now", command=self.manual_refresh)
        self.btn_manual.pack(side="left", padx=6)

        # Info labels
        info = tk.Frame(root, bg="#0d1117")
        info.pack(fill="x", padx=8, pady=(0,6))
        self.price_label = tk.Label(info, text="Price: --", font=("Segoe UI", 13), bg="#0d1117", fg="white")
        self.price_label.pack(side="left", padx=8, pady=6)
        self.rsi_label = tk.Label(info, text="RSI: --", font=("Segoe UI", 13), bg="#0d1117", fg="white")
        self.rsi_label.pack(side="left", padx=18, pady=6)
        self.signal_label = tk.Label(info, text="Signal: --", font=("Segoe UI", 13, "bold"), bg="#0d1117", fg="white")
        self.signal_label.pack(side="left", padx=18, pady=6)

        # Main frame: charts + history
        main = tk.Frame(root)
        main.pack(fill="both", expand=True, padx=8, pady=6)

        left = tk.Frame(main)
        left.pack(side="left", fill="both", expand=True)

        # Matplotlib figure & axes (manual layout to avoid tight_layout warnings)
        self.fig = plt.Figure(figsize=(9,6), dpi=100)
        self.ax_price = self.fig.add_axes([0.06, 0.38, 0.92, 0.58])
        self.ax_rsi = self.fig.add_axes([0.06, 0.08, 0.92, 0.24], sharex=self.ax_price)

        self.canvas = FigureCanvasTkAgg(self.fig, master=left)
        self.canvas.get_tk_widget().pack(fill="both", expand=True)

        right = tk.Frame(main, width=320)
        right.pack(side="right", fill="y")
        tk.Label(right, text="Signal History (latest on top)", font=("Segoe UI", 10, "bold")).pack(anchor="w", padx=6, pady=(6,0))
        self.history_box = tk.Listbox(right, width=55, height=30, bg="#0f0f0f", fg="white", font=("Consolas",10))
        self.history_box.pack(padx=6, pady=6, fill="y")

        # Poll queue periodically
        self.root.after(500, self.process_queue)

    # ----------------- Thread control -----------------
    def start_fetch(self):
        if self.fetch_thread and self.fetch_thread.is_alive():
            return
        self.stop_event.clear()
        self.fetch_thread = FetchThread(
            symbol_getter=lambda: self.combo.get(),
            interval_getter=lambda: self.interval.get(),
            refresh_getter=lambda: self.ref_entry.get(),
            out_queue=self.queue,
            stop_event=self.stop_event
        )
        self.fetch_thread.start()
        self.btn_start.config(state="disabled")
        self.btn_stop.config(state="normal")

    def stop_fetch(self):
        self.stop_event.set()
        self.btn_start.config(state="normal")
        self.btn_stop.config(state="disabled")

    def manual_refresh(self):
        # spawn a short-lived fetch thread to get one update immediately
        threading.Thread(target=self._manual_fetch_once, daemon=True).start()

    def _manual_fetch_once(self):
        try:
            symbol_key = self.combo.get()
            symbol = SYMBOLS.get(symbol_key, list(SYMBOLS.values())[0])
            interval = self.interval.get() or DEFAULT_INTERVAL
            df = fetch_klines(symbol, interval=interval, limit=CANDLES_LIMIT)
            df["RSI"] = compute_rsi(df["Close"], period=RSI_PERIOD)
            payload = {"symbol": symbol, "df": df, "price": float(df["Close"].iloc[-1]), "rsi": float(df["RSI"].iloc[-1])}
            self.queue.put(("data", payload))
        except Exception as e:
            self.queue.put(("error", str(e)))

    # ----------------- Queue processor -----------------
    def process_queue(self):
        try:
            while True:
                tag, payload = self.queue.get_nowait()
                if tag == "data":
                    self.apply_data(payload)
                elif tag == "error":
                    self.signal_label.config(text=f"Error: {payload}", bg="#ffd54f", fg="black")
        except queue.Empty:
            pass
        self.root.after(500, self.process_queue)

    # ----------------- Apply fetched data to UI -----------------
    def apply_data(self, payload):
        df = payload["df"]
        symbol = payload["symbol"]
        latest_price = payload["price"]
        latest_rsi = payload["rsi"]

        # signal
        if np.isnan(latest_rsi):
            signal = "No RSI"
            bg = "#424242"
        elif latest_rsi >= UPPER_BAND:
            signal = "SELL (Overbought)"
            bg = "#b71c1c"
        elif latest_rsi <= LOWER_BAND:
            signal = "BUY (Oversold)"
            bg = "#1b5e20"
        elif latest_rsi > MIDDLE_BAND:
            signal = "BUY (Above Mid)"
            bg = "#2e7d32"
        else:
            signal = "SELL (Below Mid)"
            bg = "#c62828"

        # update labels
        self.price_label.config(text=f"Price: {format_price(latest_price)}")
        self.rsi_label.config(text=f"RSI: {latest_rsi:.2f}" if not np.isnan(latest_rsi) else "RSI: -")
        self.signal_label.config(text=f"Signal: {signal}", bg=bg, fg="white")

        # history
        ts = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        entry = f"{ts} | {symbol} | {format_price(latest_price)} | RSI {latest_rsi:.2f} | {signal}"
        self.history.insert(0, entry)
        if len(self.history) > MAX_HISTORY:
            self.history.pop()
        self.history_box.delete(0, tk.END)
        for itm in self.history[:100]:
            self.history_box.insert(tk.END, itm)

        # plotting
        self.ax_price.clear()
        self.ax_rsi.clear()

        try:
            # mpf.plot will draw on the provided axes
            mpf.plot(df, type='candle', ax=self.ax_price, volume=False, style='charles', show_nontrading=True)
        except Exception:
            # fallback line
            self.ax_price.plot(df.index, df['Close'], color='cyan', linewidth=1.2)
            self.ax_price.set_title(f"{symbol} Price (line fallback)")

        self.ax_price.set_ylabel("Price")
        self.ax_price.grid(alpha=0.2)

        self.ax_rsi.plot(df.index, df['RSI'], color='orange', linewidth=1.0)
        self.ax_rsi.axhline(UPPER_BAND, color='red', linestyle='--', linewidth=0.9)
        self.ax_rsi.axhline(MIDDLE_BAND, color='gray', linestyle=':', linewidth=0.8)
        self.ax_rsi.axhline(LOWER_BAND, color='green', linestyle='--', linewidth=0.9)
        self.ax_rsi.set_ylim(0, 100)
        self.ax_rsi.set_ylabel("RSI")
        self.ax_rsi.grid(alpha=0.2)

        # highlight last RSI
        if not np.isnan(latest_rsi):
            col = 'red' if latest_rsi >= UPPER_BAND else ('green' if latest_rsi <= LOWER_BAND else 'white')
            self.ax_rsi.scatter(df.index[-1], latest_rsi, color=col, zorder=6, s=36)

        # manual subplot adjustment (avoid tight_layout warnings)
        self.fig.subplots_adjust(left=0.06, right=0.98, top=0.96, bottom=0.08, hspace=0.05)
        self.canvas.draw_idle()

    def on_close(self):
        self.stop_event.set()
        self.root.destroy()

# ----------------- Run -----------------
if __name__ == "__main__":
    root = tk.Tk()
    app = RSIApp(root)
    root.protocol("WM_DELETE_WINDOW", app.on_close)
    root.mainloop()
